False negative on cloud-on-k8s-e2e-tests-ocp

Since #4593, the job fails at the end during the cleanup:

```
08:01:56  2021/06/30 06:01:56 Deleting cluster...
08:01:57  level=info msg=Credentials loaded from file "/home/.gcp/osServiceAccount.json"
08:02:43  level=info msg=Stopped instance ocp4-ci-492-92qb7-worker-c-wpnx7
...
08:06:39  level=info msg=Deleted network ocp4-ci-492-92qb7-network
08:06:39  level=info msg=Time elapsed: 4m35s
08:06:39  2021/06/30 06:06:34 Removing context, user and cluster entry from kube config

08:06:39  error: cannot delete context admin, not in /root/.kube/config <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<

08:06:39  Error: exit status 1
08:06:39  exit status 1
08:06:39  make: *** [Makefile:525: run-deployer] Error 1
08:06:39  Makefile:46: recipe for target 'ci-internal' failed
08:06:39  make[1]: *** [ci-internal] Error 2
08:06:39  make[1]: Leaving directory '/var/lib/jenkins/workspace/cloud-on-k8s-e2e-tests-ocp/.ci'
08:06:39  Makefile:40: recipe for target 'ci' failed
08:06:39  make: *** [ci] Error 2
08:06:39  make: Leaving directory '/var/lib/jenkins/workspace/cloud-on-k8s-e2e-tests-ocp/.ci'
08:06:39  [Pipeline] }
08:06:39  [Pipeline] // script
08:06:39  Error when executing cleanup post condition:
```

--- a/hack/deployer/runner/ocp.go
+++ b/hack/deployer/runner/ocp.go
@@ -411,7 +411,7 @@ func (d *OcpDriver) copyKubeconfig() error {
 func (d *OcpDriver) removeKubeconfig() error {
 	if err := NewCommand("kubectl config get-contexts admin").Run(); err != nil {
 		// skip because the admin context does not exist in the kube config
-		return nil
+		return nil //nolint:nilerr
 	}
 
 	log.Printf("Removing context, user and cluster entry from kube config")

--- a/hack/deployer/runner/ocp.go
+++ b/hack/deployer/runner/ocp.go
@@ -314,7 +314,7 @@ var (
 func (d *OcpDriver) currentStatus() ClusterStatus {
 	log.Println("Checking if cluster exists...")
 
-	kubeConfig := d.kubeconfig()
+	kubeConfig := filepath.Join(d.runtimeState.ClusterStateDir, "auth", "kubeconfig")
 	if _, err := os.Stat(kubeConfig); os.IsNotExist(err) {
 		if empty, err := isEmpty(d.runtimeState.ClusterStateDir); empty && err == nil {
 			return NotFound
@@ -323,7 +323,8 @@ func (d *OcpDriver) currentStatus() ClusterStatus {
 	}
 
 	log.Println("Cluster state synced: Testing that the OpenShift cluster is alive... ")
-	alive, err := d.kubectl("version").WithoutStreaming().OutputContainsAny("Server Version")
+	cmd := "kubectl version"
+	alive, err := NewCommand(cmd).WithoutStreaming().WithVariable("KUBECONFIG", kubeConfig).OutputContainsAny("Server Version")
 
 	if !alive || err != nil { // error will be typically not nil when alive is false but let's be explicit here to avoid returning Running on a non-nil error
 		log.Printf("a cluster state dir was found in %s but the cluster is not responding to `kubectl version`: %s", d.runtimeState.ClusterStateDir, err.Error())
@@ -397,29 +398,30 @@ func (d *OcpDriver) downloadClusterState() error {
 
 func (d *OcpDriver) copyKubeconfig() error {
 	log.Printf("Copying  credentials")
+	kubeConfig := filepath.Join(d.runtimeState.ClusterStateDir, "auth", "kubeconfig")
 
 	// 1. merge or create kubeconfig
-	if err := mergeKubeconfig(d.kubeconfig()); err != nil {
+	if err := mergeKubeconfig(kubeConfig); err != nil {
 		return err
 	}
 	// 2. after merging make sure that the ocp context is in use, which is always called `admin`
-	return d.kubectl("config use-context admin").Run()
+	return NewCommand("kubectl config use-context admin").Run()
 }
 
 func (d *OcpDriver) removeKubeconfig() error {
-	if err := d.kubectl("config get-contexts admin").Run(); err != nil {
+	if err := NewCommand("kubectl config get-contexts admin").Run(); err != nil {
 		// skip because the admin context does not exist in the kube config
 		return nil
 	}
 
 	log.Printf("Removing context, user and cluster entry from kube config")
-	if err := d.kubectl("config delete-context admin").Run(); err != nil {
+	if err := NewCommand("kubectl config delete-context admin").Run(); err != nil {
 		return err
 	}
-	if err := d.kubectl("config delete-user admin").Run(); err != nil {
+	if err := NewCommand("kubectl config delete-user admin").Run(); err != nil {
 		return err
 	}
-	return d.kubectl("config delete-cluster {{.ClusterName}}").
+	return NewCommand("kubectl config delete-cluster {{.ClusterName}}").
 		AsTemplate(map[string]interface{}{"ClusterName": d.plan.ClusterName}).
 		Run()
 }
@@ -470,16 +472,4 @@ func (d *OcpDriver) baseDomain() string {
 		baseDomain = "eck-ocp.elastic.dev"
 	}
 	return baseDomain
-}
-
-// kubectl execues a kubectl command by having set the KUBECONFIG environment variable
-func (d *OcpDriver) kubectl(args string) *Command {
-	return NewCommand(fmt.Sprintf("kubectl %s", args)).
-		WithoutStreaming().
-		WithVariable("KUBECONFIG", d.kubeconfig())
-}
-
-// kubeconfig returns the kube config file path
-func (d *OcpDriver) kubeconfig() string {
-	return filepath.Join(d.runtimeState.ClusterStateDir, "auth", "kubeconfig")
 }

--- a/hack/deployer/runner/ocp.go
+++ b/hack/deployer/runner/ocp.go
@@ -407,6 +407,11 @@ func (d *OcpDriver) copyKubeconfig() error {
 }
 
 func (d *OcpDriver) removeKubeconfig() error {
+	if err := d.kubectl("config get-contexts admin").Run(); err != nil {
+		// skip because the admin context does not exist in the kube config
+		return nil
+	}
+
 	log.Printf("Removing context, user and cluster entry from kube config")
 	if err := d.kubectl("config delete-context admin").Run(); err != nil {
 		return err

--- a/hack/deployer/runner/ocp.go
+++ b/hack/deployer/runner/ocp.go
@@ -314,7 +314,7 @@ var (
 func (d *OcpDriver) currentStatus() ClusterStatus {
 	log.Println("Checking if cluster exists...")
 
-	kubeConfig := filepath.Join(d.runtimeState.ClusterStateDir, "auth", "kubeconfig")
+	kubeConfig := d.kubeconfig()
 	if _, err := os.Stat(kubeConfig); os.IsNotExist(err) {
 		if empty, err := isEmpty(d.runtimeState.ClusterStateDir); empty && err == nil {
 			return NotFound
@@ -323,8 +323,7 @@ func (d *OcpDriver) currentStatus() ClusterStatus {
 	}
 
 	log.Println("Cluster state synced: Testing that the OpenShift cluster is alive... ")
-	cmd := "kubectl version"
-	alive, err := NewCommand(cmd).WithoutStreaming().WithVariable("KUBECONFIG", kubeConfig).OutputContainsAny("Server Version")
+	alive, err := d.kubectl("version").WithoutStreaming().OutputContainsAny("Server Version")
 
 	if !alive || err != nil { // error will be typically not nil when alive is false but let's be explicit here to avoid returning Running on a non-nil error
 		log.Printf("a cluster state dir was found in %s but the cluster is not responding to `kubectl version`: %s", d.runtimeState.ClusterStateDir, err.Error())
@@ -398,25 +397,24 @@ func (d *OcpDriver) downloadClusterState() error {
 
 func (d *OcpDriver) copyKubeconfig() error {
 	log.Printf("Copying  credentials")
-	kubeConfig := filepath.Join(d.runtimeState.ClusterStateDir, "auth", "kubeconfig")
 
 	// 1. merge or create kubeconfig
-	if err := mergeKubeconfig(kubeConfig); err != nil {
+	if err := mergeKubeconfig(d.kubeconfig()); err != nil {
 		return err
 	}
 	// 2. after merging make sure that the ocp context is in use, which is always called `admin`
-	return NewCommand("kubectl config use-context admin").Run()
+	return d.kubectl("config use-context admin").Run()
 }
 
 func (d *OcpDriver) removeKubeconfig() error {
 	log.Printf("Removing context, user and cluster entry from kube config")
-	if err := NewCommand("kubectl config delete-context admin").Run(); err != nil {
+	if err := d.kubectl("config delete-context admin").Run(); err != nil {
 		return err
 	}
-	if err := NewCommand("kubectl config delete-user admin").Run(); err != nil {
+	if err := d.kubectl("config delete-user admin").Run(); err != nil {
 		return err
 	}
-	return NewCommand("kubectl config delete-cluster {{.ClusterName}}").
+	return d.kubectl("config delete-cluster {{.ClusterName}}").
 		AsTemplate(map[string]interface{}{"ClusterName": d.plan.ClusterName}).
 		Run()
 }
@@ -467,4 +465,16 @@ func (d *OcpDriver) baseDomain() string {
 		baseDomain = "eck-ocp.elastic.dev"
 	}
 	return baseDomain
+}
+
+// kubectl execues a kubectl command by having set the KUBECONFIG environment variable
+func (d *OcpDriver) kubectl(args string) *Command {
+	return NewCommand(fmt.Sprintf("kubectl %s", args)).
+		WithoutStreaming().
+		WithVariable("KUBECONFIG", d.kubeconfig())
+}
+
+// kubeconfig returns the kube config file path
+func (d *OcpDriver) kubeconfig() string {
+	return filepath.Join(d.runtimeState.ClusterStateDir, "auth", "kubeconfig")
 }

